text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1740 @@
+}
+```
+
+It is possible to do this in a backwards compatible manner because the token `:`
+is not in the follow set of `pat` fragments. This means that when you write
+
+```rust
+macro_rules! test {
+    ($p:pat : u32) => {}
+}
+```
+
+The compiler will complain that:
+
+```rust
+error: `$p:pat` is followed by `:`, which is not allowed for `pat` fragments
+ --> src/main.rs:2:12
+  |
+2 |     ($p:pat : u32) => {}
+  |             ^
+```
+
+### Let bindings
+
+Before this RFC when you wrote something like:
+
+```rust
+let quux: u8 = 42;
+    ^^^^
+```
+
+The underlined part was the pattern, but the typing annotation `: u8` to
+the right was *not* part of the pattern. With this RFC, we unify the language
+and we can now say that everything after `let` and before `=` is the pattern:
+
+```rust
+let quux: u8 = 42;
+    ^^^^^^^^
+    Pattern!
+```
+
+Another implication of introducing type ascription in pattern contexts is that
+that you may say things like:
+
+```rust
+let [alpha: u8, beta, gamma] = [1, 2, 3];
+
+let (alpha: u8, beta: i16, gamma: bool) = (1, -2, true);
+```
+
+### Function definitions
+
+Another change that comes with this RFC is that the fundamental unit of a
+function parameter also becomes a pattern. Thus when you write things like:
+
+```rust
+fn foo(alpha: usize, beta: bool) { ... }
+       ^^^^^^^^^^^^  ^^^^^^^^^^^
+       Pattern       Pattern
+```
+
+the underlined parts are the patterns while before this RFC, the patterns were:
+
+```rust
+fn foo(alpha: usize, beta: bool) { ... }
+       ^^^^^         ^^^^
+       Pattern       Pattern
+```
+
+Indeed, with the change to function definitions, it becomes
+*syntactically valid* to write:
+
+```rust
+fn foo(alpha, beta) -> usize { .. }
+```
+
+Nevertheless, as we want to avoid introducing global type inference to the
+language, the type checker will prevent this from compiling and will emit an
+error which *may* look like:
+
+```rust
+error[E0282]: type annotations needed
+ --> src/main.rs:1:9
+  |
+1 |     fn foo(alpha, beta) -> usize { .. }
+  |            ^^^^^  ^^^^
+  |            |      |
+  |            |______|
+  |            |
+  |            The following patterns do not have a fully determined type.
+  |            help: Write type annotations on the following patterns:
+  |
+  |                alpha: usize
+  |                beta: bool
+  |
+
+error: aborting due to previous error
+```
+
+This also gives the compiler an opportunity to tell you what the types are
+if it so happens that you need this help. Providing the user with this
+type information is not mandatory and is instead up to the implementation
+of the compiler.
+
+However, in some cases, you can determine the type from the pattern alone.
+Therefore, when the type is fully determined you may omit the type ascription.
+Thus, the following definitions are legal:
+
+```rust
+fn foo<T>(Wrapping(value: T)) -> usize { ... }
+
+fn bar(Wrapping(value: usize)) -> usize { ... }
```

Comment:
While I really like this feature, it might be worth scoping it down initially to avoid needing to specify all the details.  For example, there are no examples of lifetimes, thus far.  (And I think it'd work fine, but there are a bunch of questions around things like lifetime elision if you have `Foo { x: x: &i32, y: y: &i32 }` and it's `Foo<'_>`.)

Maybe start with just the "no generics" case?  (Like `Quux` below.)

Reply:
...or you could just define the full unification algorithm :rofl: